core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```
@@ -225,10 +233,9 @@ abstract class AbstractFetcherThread(name: String,
       }
     }
 
-    if (partitionsWithError.nonEmpty) {
+    if (partitionsWithError.nonEmpty)
       debug("handling partitions with error for %s".format(partitionsWithError))
-      handlePartitionsWithErrors(partitionsWithError)
-    }
+    handlePartitionsWithErrors(partitionsWithError)
```

Instead of calling handlePartitionsWithErrors() multiple times, it's probably better for ResultWithPartitions to return a set of error partitions and the corresponding exception. Then, we can collect all error partitions and in the end, log the error with the exception, call partitionStates.moveToEnd() and call the error handler here in one place.

---

I have updated the code to remove invocation of `partitionStates.moveToEnd()`. It can be removed because the current implementation of handlePartitionsWithErrors essentially overwrites the effect of partitionStates.moveToEnd(). And going forward we can call partitionStates.moveToEnd() in handlePartitionsWithErrors() if needed.

But I am not sure why we need to return the corresponding exception in ResultWithPartitions. Currently handlePartitionsWithErrors() does not depend on the actual exception. By returning the corresponding exception in ResultWithPartitions, do you want to also centralize the error logging? I understand that we want to centralising side-effects (e.g. only call handlePartitionsWithErrors in AbstractFetcherThread) to make the code easier to reason about. But it seems OK to log error simply when it happens, right?